In [1]:
import sys
import os
sys.path.append(os.path.abspath('..'))

import numpy as np
import polars as pl

from src.data.loader import SAMLDDataLoader

from src.analysis.transaction_eda import SAMLDTransactionEDA

In [2]:
### load standardized transactions
# create loader instance
loader = SAMLDDataLoader()

# lazily scan transactions from interim Parquet dataset
transactions = loader.scan_interim()

# create transaction EDA instance
transaction_eda = SAMLDTransactionEDA(
    transactions = transactions
)

In [3]:
### dataset and target overview
# dataset overview
transaction_overview = transaction_eda.get_transaction_overview()

# target overview
label_distribution = transaction_eda.get_label_distribution()

display(transaction_overview)
display(label_distribution)

transaction_count,unique_transaction_id_count,unique_sender_account,unique_receiver_account,normal_transaction_count,laundering_transaction_count,minimum_timestamp,maximum_timestamp,laundering_rate
u64,u64,u64,u64,u64,u64,datetime[μs],datetime[μs],f64
9504852,9504852,292715,652266,9494979,9873,2022-10-07 10:35:19,2023-08-23 10:57:12,0.001039


is_laundering,label,transaction_count,transaction_share
i8,str,u64,f64
0,"""normal""",9494979,0.998961
1,"""laundering""",9873,0.001039


In [4]:
### daily transaction activity
# daily activity
daily_activity = transaction_eda.get_daily_activity()

display(daily_activity)

transaction_date,transaction_count,laundering_transaction_count,active_sender_count,active_receiver_account,laundering_rate
date,u64,u64,u64,u64,f64
2022-10-07,20892,27,9110,11111,0.001292
2022-10-08,29023,29,10072,13863,0.000999
2022-10-09,28880,32,10300,13570,0.001108
2022-10-10,27345,32,11123,14632,0.00117
2022-10-11,28704,30,11248,14782,0.001045
…,…,…,…,…,…
2023-08-19,28206,29,10947,14565,0.001028
2023-08-20,30745,36,11012,14764,0.001171
2023-08-21,29094,48,12009,15526,0.00165


In [5]:
### categorical summaries
# payment type categorical summary
payment_type_summary = transaction_eda.get_categorical_summary(
    column_name = 'payment_type'
)

# payment currency categorical summary
payment_currency_summary = transaction_eda.get_categorical_summary(
    column_name = 'payment_currency'
)

# received currency categorical summary
received_currency_summary = transaction_eda.get_categorical_summary(
    column_name = 'received_currency'
)

# sender bank location categorical summary
sender_location_summary = transaction_eda.get_categorical_summary(
    column_name = 'sender_bank_location'
)

# receiver bank location categorical summary
receiver_location_summary = transaction_eda.get_categorical_summary(
    column_name = 'receiver_bank_location'
)

with pl.Config(tbl_rows = -1):
    display(payment_type_summary)
    display(payment_currency_summary)
    display(received_currency_summary)
    display(sender_location_summary)
    display(receiver_location_summary)

payment_type,transaction_count,laundering_transaction_count,transaction_share,laundering_rate
str,u64,u64,f64,f64
"""Credit card""",2012909,1136,0.211777,0.000564
"""Debit card""",2012103,1124,0.211692,0.000559
"""Cheque""",2011419,1087,0.21162,0.00054
"""ACH""",2008807,1159,0.211345,0.000577
"""Cross-border""",933931,2628,0.098258,0.002814
"""Cash Withdrawal""",300477,1334,0.031613,0.00444
"""Cash Deposit""",225206,1405,0.023694,0.006239


payment_currency,transaction_count,laundering_transaction_count,transaction_share,laundering_rate
str,u64,u64,f64,f64
"""UK pounds""",9099293,8830,0.957331,0.00097
"""Euro""",117164,260,0.012327,0.002219
"""Turkish lira""",27996,73,0.002945,0.002608
"""Swiss franc""",27492,82,0.002892,0.002983
"""Dirham""",27263,89,0.002868,0.003264
"""Pakistani rupee""",27196,62,0.002861,0.00228
"""Naira""",27143,62,0.002856,0.002284
"""US dollar""",26061,60,0.002742,0.002302
"""Yen""",25562,65,0.002689,0.002543


received_currency,transaction_count,laundering_transaction_count,transaction_share,laundering_rate
str,u64,u64,f64,f64
"""UK pounds""",8783655,6919,0.924123,0.000788
"""Euro""",231911,716,0.024399,0.003087
"""Pakistani rupee""",45993,204,0.004839,0.004435
"""Yen""",45814,118,0.00482,0.002576
"""Moroccan dirham""",45748,285,0.004813,0.00623
"""Albanian lek""",45736,262,0.004812,0.005729
"""Mexican Peso""",45255,235,0.004761,0.005193
"""Naira""",45046,288,0.004739,0.006393
"""Indian rupee""",43757,172,0.004604,0.003931


sender_bank_location,transaction_count,laundering_transaction_count,transaction_share,laundering_rate
str,u64,u64,f64,f64
"""UK""",9183088,9253,0.966147,0.001008
"""Turkey""",20902,27,0.002199,0.001292
"""Switzerland""",20503,41,0.002157,0.002
"""Pakistan""",20346,25,0.002141,0.001229
"""UAE""",20081,34,0.002113,0.001693
"""Nigeria""",20027,25,0.002107,0.001248
"""Spain""",19391,27,0.00204,0.001392
"""Germany""",19259,43,0.002026,0.002233
"""USA""",19027,36,0.002002,0.001892


receiver_bank_location,transaction_count,laundering_transaction_count,transaction_share,laundering_rate
str,u64,u64,f64,f64
"""UK""",8864634,7308,0.932643,0.000824
"""Pakistan""",38980,169,0.004101,0.004336
"""Austria""",38844,154,0.004087,0.003965
"""Albania""",38783,235,0.00408,0.006059
"""Japan""",38729,83,0.004075,0.002143
"""Morocco""",38704,253,0.004072,0.006537
"""Germany""",38540,102,0.004055,0.002647
"""France""",38512,82,0.004052,0.002129
"""Mexico""",38385,197,0.004038,0.005132


In [6]:
### currency pair & amount summaries
# currency pair summary
currency_pair_summary = transaction_eda.get_currency_pair_summary()

# amount summary
amount_summary = transaction_eda.get_amount_summary()

with pl.Config(tbl_rows = -1):
    display(currency_pair_summary)
    display(amount_summary)

payment_currency,received_currency,transaction_count,laundering_transaction_count,same_currency,transaction_share,laundering_rate
str,str,u64,u64,bool,f64,f64
"""UK pounds""","""UK pounds""",8414112,6171,true,0.885244,0.000733
"""UK pounds""","""Euro""",220504,625,false,0.023199,0.002834
"""Euro""","""UK pounds""",106608,186,false,0.011216,0.001745
"""UK pounds""","""Pakistani rupee""",43895,187,false,0.004618,0.00426
"""UK pounds""","""Yen""",43517,101,false,0.004578,0.002321
"""UK pounds""","""Moroccan dirham""",43461,261,false,0.004573,0.006005
"""UK pounds""","""Mexican Peso""",43174,220,false,0.004542,0.005096
"""UK pounds""","""Albanian lek""",43040,232,false,0.004528,0.00539
"""UK pounds""","""Naira""",42696,261,false,0.004492,0.006113


payment_currency,is_laundering,transaction_count,minimum_amount,amount_q25,median_amount,amount_q75,amount_q95,amount_q99,maximum_amount,mean_amount
str,i8,u64,f64,f64,f64,f64,f64,f64,f64,f64
"""Albanian lek""",0,24714,18.82,3903.935,7154.855,11633.075,23832.511,48922.2801,935820.72,9615.598072
"""Albanian lek""",1,64,2800.75,4749.345,6506.99,11294.895,34732.939205,52116.954846,75683.499559,10711.803623
"""Dirham""",0,27174,17.17,4593.8375,7161.275,11694.1625,23679.2715,49462.5739,956193.0,9907.217582
"""Dirham""",1,89,2259.18,4245.38,6289.9,9798.31,37250.534444,58730.145957,62867.858744,10902.371375
"""Euro""",0,116904,9.15,4972.83,7642.735,12156.18,25429.531,53604.7423,995675.88,10523.269269
"""Euro""",1,260,1250.41,4470.5075,6497.501465,11887.8375,66535.648256,154557.239406,294167.63,16758.714386
"""Indian rupee""",0,24589,13.41,4646.09,7323.48,11936.55,24032.4,55993.2252,985678.03,10217.446434
"""Indian rupee""",1,68,2618.8,4088.455,6020.887746,9534.081785,23943.53339,66606.854691,67635.42,9736.177476
"""Mexican Peso""",0,24784,5.9,4632.9925,7395.62,11741.6075,25546.058,50473.1953,992045.52,10212.466525


In [7]:
### laundering typology summary
# laundering typology
laundering_typology_summary = transaction_eda.get_typology_summary(
    label = 1
)

# legitimate typology
normal_typology_summary = transaction_eda.get_typology_summary(
    label = 0
)

with pl.Config(tbl_rows = -1):
    display(laundering_typology_summary)
    display(normal_typology_summary)

is_laundering,laundering_type,transaction_count,share_within_label
i8,str,u64,f64
1,"""Structuring""",1870,0.189405
1,"""Stacked Bipartite""",506,0.051251
1,"""Smurfing""",932,0.094399
1,"""Single_large""",250,0.025322
1,"""Scatter-Gather""",338,0.034235
1,"""Over-Invoicing""",54,0.005469
1,"""Layered_Fan_Out""",529,0.05358
1,"""Layered_Fan_In""",656,0.066444
1,"""Gather-Scatter""",354,0.035855


is_laundering,laundering_type,transaction_count,share_within_label
i8,str,u64,f64
0,"""Normal_single_large""",20641,0.002174
0,"""Normal_Small_Fan_Out""",3477717,0.366269
0,"""Normal_Plus_Mutual""",155041,0.016329
0,"""Normal_Periodical""",210526,0.022172
0,"""Normal_Mutual""",125335,0.0132
0,"""Normal_Group""",528351,0.055645
0,"""Normal_Foward""",42031,0.004427
0,"""Normal_Fan_Out""",2302220,0.242467
0,"""Normal_Fan_In""",2104285,0.221621


In [8]:
### account identity overview
# account overview
account_overview = transaction_eda.get_account_overview()

display(account_overview)

unique_account_count,suspicious_account_count,sender_and_receiver_count,sender_only_count,receiver_only_count,multi_location_account_count,suspicious_account_rate
u64,u64,u64,u64,u64,u64,f64
855460,7902,89521,203194,562745,6507,0.009237


In [9]:
### multi location account summary
# multi bank location account count
multi_location_account_count = int(
    account_overview['multi_location_account_count'][0]
)

# check if multi location account count is non-negative
if multi_location_account_count > 0:
    multi_location_accounts = transaction_eda.get_multi_location_accounts(
        limit = 20
    )

    display(multi_location_accounts)
else:
    print(
        'No accounts are associated with multiple bank locations.'
    )

# total multi_location accounts
if multi_location_account_count > 0:
    max_int32 = np.iinfo(np.int32).max

    max_multi_location_accounts = transaction_eda.get_multi_location_accounts(
        limit = max_int32
    )

    print(
        f'Total multi-location accounts: {max_multi_location_accounts.shape[0]}'
    )

account,bank_location_count,bank_locations,account_event_count
str,u32,list[str],u64
"""7823743088""",5,"[""Austria"", ""India"", … ""USA""]",810
"""6636772831""",5,"[""Nigeria"", ""Spain"", … ""UK""]",777
"""6517474457""",5,"[""Austria"", ""India"", … ""UK""]",723
"""5708009014""",5,"[""Albania"", ""Turkey"", … ""USA""]",710
"""1168600456""",5,"[""India"", ""Netherlands"", … ""UK""]",585
…,…,…,…
"""4035147493""",4,"[""Austria"", ""Italy"", … ""UK""]",739
"""453377489""",4,"[""Japan"", ""Morocco"", … ""UK""]",733
"""2384413935""",4,"[""Morocco"", ""Netherlands"", … ""UK""]",729


Total multi-location accounts: 6507


In [10]:
### graph-size and duplicate summaries
# directed pair summary
directed_pair_summary = transaction_eda.get_directed_pair_summary()

# duplicate summary
duplicate_summary = transaction_eda.get_duplicate_summary()

display(directed_pair_summary)
display(duplicate_summary)

directed_pair_count,repeated_directed_pair_count,maximum_transactions_per_pair,mean_transactions_per_pair,repeated_directed_pair_rate
u64,u64,u64,f64,f64
887497,781631,84,10.709729,0.880714


duplicate_group_count,rows_in_duplicate_groups,duplicate_excess_row_count,maximum_multiplicity
u64,u64,u64,u64
0,0,0,0


**Design Decisions**

Translating transaction-level EDA outputs into graph-design decisions:

- 887k directed account pairs out of 9.5m total transactions
    - Aggregate transactions by sender-receiver pair
- 88.07% of directed pairs repeat
    - Use transaction count as the first edge weight
- No exact duplicates
    - No deduplication step is required
- 13 currencies and strongly skewed amounts
    - Should not be summing raw amounts across currencies
- 6507 multi-location accounts
    - Keep one node per account
- 0.1039% laundering transactions
    - Use ranking metrics
- 17 laundering typologies
    - Report results separately by typology
- Final date contains only 8.4k transactions
    - Exclude the incomplete date from test evaluation